New doc: How Radicle works (#194)
radicle-dev/radicle-docs: docs/understanding-radicle/how-radicle-works.md

infrastructure and incentives for open-source work.

If you're curious about that global vision, we recommend reading [_Towards decentralized code
collaboration_](https://radicle.mirror.xyz/Qz4Rr0R54KLPurqqswsQ2tfRtBs9MOYpw3dlY9VjyQ8) and [_Introducing
RAD_](https://radicle.mirror.xyz/CgcHpSXUlPvwMVaUVVaJ7r8bIJI2BOKOytaI9-nO9oY), which details the governance token used
by the DAO.

However, for the sake of this document, we'll focus on the user-facing components of this ecosystem &mdash; what you can
interact with, as a user, to securely collaborate on software over a peer-to-peer network built on Git.

## Key components of the Radicle ecosystem

There are a few important open-source protocols, concepts, projects, and assets that integrate to make [Radicle's
vision](https://radicle.mirror.xyz/Qz4Rr0R54KLPurqqswsQ2tfRtBs9MOYpw3dlY9VjyQ8) for decentralized code collaboration
possible.

- [**Identity**](#radicle-identity): A cryptographically-signed JSON document that identifies a device or a project,
  stored alongside standard Git data when hosted, cloned, and replicated.
- [**Network**](#radicle-network): The people who use Radicle tooling and the projects they collaborate on,

>   which is supported by seed nodes, which provide a data availability layer through a monorepo of projects and
>   identities.

Reviewer: The monorepo is an implementation detail, which might go away soon.

```suggestion
  which is supported by seed nodes, which provide a data availability layer.
```

- [**Peer-to-peer protocol**](#peer-to-peer-protocol): The protocol for disseminating Git repositories via gossip-based
  replication, enabling the hosting and sharing of projects without knowing where it's physically stored within the
  Radicle network.
- [**Command-line tooling (`rad`)**](#command-line-tooling-rad): The core tool for managing your code, project, and
  identity.
- [**Web app**](#web-app): The web-based interface for discovering, viewing, and sharing Radicle projects.

> - [**Open-source funding (Drips)**](#open-source-funding-drips): An Ethereum protocol for continuously stream funds to

Reviewer: ```suggestion
- [**Open-source funding (Drips)**](#open-source-funding-drips): An Ethereum protocol for continuously streaming funds to
```

  others, which will be integrated into the [web app](#web-app) as new value flows for developers to sustain their FOSS
  work.

## Radicle identity

Everything Radicle is trying to build around decentralized code collaboration starts with the concept of identity. In
Radicle, your identity is a *portable* JSON document that defines a user or a project on the Radicle network.

> The most important part is the hashed content identifier—the **peer ID**—which identifies it uniquely wherever it's

Reviewer: I think this is not quite correct. The content identifier (URN) is not the same as the Peer ID (which designates a device). Maybe @cloudhead can jump in to clarify this?

Reviewer 2: Yeah. Peer ID represents a device.

Reviewer 3: To elaborate, the Peer ID is a hashed Ed25519 public key. This is the cryptographic identifier for a device and is tied to the signing key that the person uses to attest to their actions within the protocol. Others than can then verify these actions using the public key.

}
```

An identity can also be extended with additional metadata as needed by the client or interface that's interacting with
it.

When you create a user identity, it's signed with an Ed25519 keypair that ensures only you can change the identity and
push changes to the Radicle network on behalf of your peer ID.

When you create a project, your identity is added as a **delegate**—more on those in a moment.

### Why Radicle starts with identity

At their core, Git repositories are just commit histories. Eventually, these commit histories develop into complex trees
of folders, files, and code.

But if you take two Git repositories at face value, there's no genuine way of comparing them. You can look at the commit
logs and diff individual files, but there's no way of knowing, just through the data that is stored in Git, whether
project A and project B are meant to converge someday. Are their histories and `HEAD` meant to be the same? Can you tell which repository is meant to be canonical?

All this uncertainty happens because source code is *portable*—it can be cloned, edited, and hosted somewhere else
without the project's maintainer knowing about it.

> Centralized code collaboration platforms like GitHub and GitLab create also create identities for the projects they

Reviewer: ```suggestion
Centralized code collaboration platforms like GitHub and GitLab also create identities for the projects they
```

host. If you create two identical repositories on these platforms, their identities are `github.com/bob/project` and
`gitlab.com/bob/project`. You might know they're the same, but to someone else, there's no guarantee that the projects
are the same despite seemingly sharing a `bob/project` identity.

With centralized platforms, the code remains portable, but the *identity is non-portable*.

> A portable Radicle identity solves this problem because you always own your Radicle identity. Your identity might be replicated across the Radicle network, but because your key signs it, it's always yours. Your project may be replicated across the Radicle network, but because your identity is a delegate, only you can change the canonical remote/branch.

Reviewer: Not sure we have really solved the problem of what is canonical within a project scope yet. In the end it still might be a per-project convention. 🤷  So maybe it should be something like:

```suggestion
A portable Radicle identity solves this problem because you always own your Radicle identity. Your identity might be replicated across the Radicle network, but because your key signs it, it's always yours. Your project may be replicated across the Radicle network, but because your identity is a delegate, only you can change your own remote/branch.
```

Radicle identities are perfect for a peer-to-peer protocol because *the data is shipped with the source code, but not in
the source code*.

### Delegates

Aside from is portable identity, the most important part of a project is the delegates.

When you create a project, you add your identity as a delegate. Every time you push new changes to the project, they're
signed with your Radicle identity and your key.

When a project has multiple contributors, all of whom may be pushing code and patches to their individual remotes, the

> delegate's remote is considered the "canonical" version of the project's code base.

Reviewer: This is still a bit unclear to me. If you have multiple delegates, they still all have their own branches, which may well be diverging. I think we might leave it up to projects to decide how they determine canonicity between multiple delegates. For example it might be just quorum on the most recent commit in a branch.

If we have a project with 2 delegates and their respective default branches `main` (commits from bottom to top, top being the newest):
```
commit 3                 commit 5
commit 2                 commit 2
commit 1                 commit 1
delegate A `main` |  delegate B `main`
```

Quorum here is at commit 2.

But there might also be other mechanisms that projects choose to determine what is the agreed upon latest state.

Author: Thanks for the information and clarity—I really like the explanation you gave here, so I went ahead and adapted it for the doc itself!

I wrote a large chunk of this section all the way back in March or April of this year, so I think a *lot* has changed around the way identity is implemented across the network.

## Radicle network

The Radicle network is the web of connections between users (via their identities, devices, and projects) and protocols.
No one owns this network, and it's all built on open-source code.

As we continue developing the [peer-to-peer protocol](#peer-to-peer-protocol) that enables decentralization via Git
gossip and replication, all your interactions with the Radicle network happen via seed nodes.

### Seed nodes

Seed nodes are the data availability layer for the Radicle network, designed to ensure that even if everyone in your
peer-to-peer network is offline, you can always retrieve Radicle project data from a seed node. Seed nodes store your
identity, but not your private keys, which means they can't take action on your behalf.

> Seed nodes are based on fully [open-source

Reviewer: This link is broken.

code](https://app.radicle.network/alt-clients.radicle.eth/rad:git:hnrkk9c4zt9thuxhwi1ukxqcrs5tmhbtcsony/tree/a19cdcb846a8360d8f55c10aac8175368881a6ee),
and Radicle currently maintains three seed nodes &mdash; willow.radicle.garden, pine.radicle.garden, and

> maple.radicle.garden &mdash; with identical functionality.

Reviewer: We could also link to the seeds: https://app.radicle.xyz/seeds/maple.radicle.garden

> Like each person using Radicle, seed nodes store data in a monorepo, which organizes Radicle identities, project

Reviewer: Again, the monorepo is an implementation detail that might go away.

metadata, and multiple remotes from anyone who has collaborated on a specific project. The data looks a little bit like
this:

```jsx

> namespaces/
>   refs/

Reviewer: ```suggestion
refs/
  namespaces/
```

    <project-one>/
      refs/
        heads/
        remotes/
          alice/
          eve/
    <project-two>/
      refs/
        heads/
        remotes/
          bob/
          tom/
          alice/
          eve/
    ...
```

> Seed nodes have three primary components.

Reviewer: ```suggestion
Seed nodes have two primary components.
```

> - The **HTTP API** (`radicle-http-api`), which allows for browsing Radicle projects on the web. It's the foundation of

Reviewer: ```suggestion
- The **HTTP API** (`radicle-http-api`), which allows for browsing Radicle projects on the web. It's the back-end of
```

  the [Radicle web app](https://app.radicle.xyz/), but the API is entirely [open
  source](https://github.com/radicle-dev/radicle-interface) for anyone else to develop additional clients.
- The **Git server** (`radicle-git-server`), which acts as a bridge between the seed node's Git server and those who use
  it, replicating Git data and Radicle identities.

The Git server is the primary mode of replication at the moment, and it behaves in a push-pull fashion, just like
running `git push` or `git pull` to a repository you have with GitHub or any other Git provider. A lot happens behind
the scenes, like pushing identities, validating signatures, and replicating additional data like peers and sub-trees.

We could stop here, because seed nodes are already decentralized and open source. If you want to run a private Radicle
network, you can [host your own seed node](https://github.com/radicle-dev/radicle-client-services) for true peer-to-peer
collaboration.

But, the push-pull relationship between a local machine, seed node, and collaborators means that the project's
location—which seed node contains the most up-to-date state—is a prominent part of the user experience. You *have* to

> know and remember which seed note you're working with.

Reviewer: ```suggestion
know and remember which seed node you're working with.
```

---
id: how-radicle-works
title: How Radicle works
---

import Highlight from '@site/src/components/Highlight'

At its most global, Radicle is a decentralized autonomous organization (DAO) solving challenges around sovereign code

> infrastructure and incentives for open-source work.

Reviewer 2: I wouldn't characterize it this way, I think Radicle is a software stack, product and/or technology first and foremost. The Radicle DAO is the organization behind it.

If you're curious about that global vision, we recommend reading [_Towards decentralized code
collaboration_](https://radicle.mirror.xyz/Qz4Rr0R54KLPurqqswsQ2tfRtBs9MOYpw3dlY9VjyQ8) and [_Introducing
RAD_](https://radicle.mirror.xyz/CgcHpSXUlPvwMVaUVVaJ7r8bIJI2BOKOytaI9-nO9oY), which details the governance token used
by the DAO.

However, for the sake of this document, we'll focus on the user-facing components of this ecosystem &mdash; what you can
interact with, as a user, to securely collaborate on software over a peer-to-peer network built on Git.

## Key components of the Radicle ecosystem

There are a few important open-source protocols, concepts, projects, and assets that integrate to make [Radicle's
vision](https://radicle.mirror.xyz/Qz4Rr0R54KLPurqqswsQ2tfRtBs9MOYpw3dlY9VjyQ8) for decentralized code collaboration
possible.

> - [**Identity**](#radicle-identity): A cryptographically-signed JSON document that identifies a device or a project,

Reviewer 3: ```suggestion
- [**Identity**](#radicle-identity): A cryptographically-signed JSON document that identifies a person or a project,
```

  stored alongside standard Git data when hosted, cloned, and replicated.
- [**Network**](#radicle-network): The people who use Radicle tooling and the projects they collaborate on, supported by

>   seed nodes, which provide a data availability layer for projects and identities.

Reviewer 3: Previously you said a project was an identity.

  to others, which will be integrated into the [web app](#web-app) as new value flows for developers to sustain their
  FOSS work.

## Radicle identity

Everything Radicle is trying to build around decentralized code collaboration starts with the concept of identity. In
Radicle, your identity is a *portable* JSON document that defines a user or a project on the Radicle network.

The most important part is the hashed content identifier—the **peer ID**—which identifies it uniquely wherever it's
stored or discovered on the Radicle network. Each Radicle identity also has a few required metadata fields, which
determine whether it's a user or project identity:

```rust
struct User {
    /// A short name (nickname, handle), without any prefix such as the `@`
    /// character
    name: String,
}

struct Project {
    /// A short name
    name: String,

    /// A slightly longer description (should fit in a headline)
    description: Option<String>,

    /// The default branch. "master" is assumed for git repositories if
    /// unspecified.
    default_branch: Option<String>,
}
```

> An identity can also be extended with additional metadata as needed by the client or interface that's interacting with

Reviewer 3: What's the plan for more tutorial-like documentation?

To me, it would make sense to link to "How to create your first person and project" above and then "How to extend your metadata" here.

Author: I'll probably start planning as soon as this PR gets merged :)

Is there any current practical use for extending metadata right now? This sentence is pulled from an old @cloudhead presentation on identity, so it might not be super relevant any more. Are issues considered metadata for a project's identity?

Reviewer 3: > Is there any current practical use for extending metadata right now? This sentence is pulled from an old @cloudhead presentation on identity, so it might not be super relevant any more. Are issues considered metadata for a project's identity?

The idea for extending metadata is to allow application developers to add their own data about a `Project` or `Person` and build on top of those extensions. For example, they could add an Ethereum address (which I think was done at some point?). So TL;DR, there are infinite practical uses, as long as you can imagine it :)

it.

When you create a user identity, it's signed with an Ed25519 keypair that ensures only you can change the identity and
push changes to the Radicle network on behalf of your peer ID.

When you create a project, your identity is added as a **delegate**—more on those in a moment.

### Why Radicle starts with identity

> At their core, Git repositories are just commit histories. Eventually, these commit histories develop into complex trees

Reviewer 3: I'm not sure I agree with this line of reasoning. In the case of a project, you correctly point out that `github.com/bob/project` identifies a project. This is a stable way of referring to a project. The centralised aspect means that GitHub can ensure that there aren't two projects called `project` and so there's no mistaking you're looking at `bob/project` when `bob` shares their project.

So, how is this solved in a decentralised setting? You and I could create `project` and now they're named the same thing. We need a stable identity for our respective `project`s. We look towards [Zooko's triangle](https://en.wikipedia.org/wiki/Zooko%27s_triangle) and have a decentralized identifier (but not human-meaningful :disappointed:). The URN becomes the unique way of identifying a project.

Similarly, for personal identity, we are used to email addresses identifying ourselves. They're also centralised. So what do we do? I create a `Person` that I can share with others that uniquely identifies me as being me.

Note that the uniqueness comes from our Ed25519 keys being unique (in end-of-the-universe-computation theory :smile:).

Reviewer 3: > Identity is stored and shipped within Git data, not a seed node or peer

i.e. all data is local-first

> All actions within a project are cryptographically verified

Yes!

> URNs add uniqueness to naming conventions/schemes that are otherwise centralized

Sure, that implies stable identifiers though :) Uniqueness and searching are two sides of the same coin here

Author: Thanks! Gave this section a tweak, mostly to add these items as bullets and remove much of the old content around uniqueness.

of folders, files, and code.

But if you take two Git repositories at face value, there's no genuine way of comparing them. You can look at the commit
logs and diff individual files, but there's no way of knowing, just through the data that is stored in Git, whether
project A and project B are meant to converge someday. Are their histories and `HEAD` meant to be the same? Can you tell which repository is meant to be canonical?

All this uncertainty happens because source code is *portable*—it can be cloned, edited, and hosted somewhere else
without the project's maintainer knowing about it.

Centralized code collaboration platforms like GitHub and GitLab also create identities for the projects they host. If
you create two identical repositories on these platforms, their identities are `github.com/bob/project` and
`gitlab.com/bob/project`. You might know they're the same, but to someone else, there's no guarantee that the projects

> are the same despite seemingly sharing a `bob/project` identity.

Reviewer 2: they don't share an identity, since they are on different platforms.

With centralized platforms, the code remains portable, but the *identity is non-portable*.

A portable Radicle identity solves this problem because you always own your Radicle identity. Your identity might be
replicated across the Radicle network, but because your key signs it, it's always yours. Your project may be replicated
across the Radicle network, but because your identity is a delegate, only you can change your own remote/branch.

Radicle identities are perfect for a peer-to-peer protocol because *the data is shipped with the source code, but not in
the source code*.

### Delegates

> Aside from is portable identity, the most important part of a project is the delegates.

Reviewer 3: "Aside from is portable identity", I'm not sure this makes sense :)

When you create a project, you add your identity as a delegate. Every time you push new changes to the project, they're
signed with your Radicle identity and your key.

Those involved in a Radicle project can choose to manage delegates and decide on canonical versions however they'd like.
If you're the only contributor to a project, and you're the delegate, then you'd likely consider the tip of your
`main`/`master` branch to be the canonical version.

> When a project has multiple contributors and/or delegates, their remotes diverge as they individually add and push code

Reviewer 3: Does the clients implementation support multiple delegates? cc @cloudhead 

If not, it might be worth dropping this.

Reviewer 2: There's no way in the CLI to add multiple delegates atm.

  it, replicating Git data and Radicle identities.

The Git server is the primary mode of replication at the moment, and it behaves in a push-pull fashion, just like
running `git push` or `git pull` to a repository you have with GitHub or any other Git provider. A lot happens behind
the scenes, like pushing identities, validating signatures, and replicating additional data like peers and sub-trees.

We could stop here, because seed nodes are already decentralized and open source. If you want to run a private Radicle
network, you can [host your own seed node](https://github.com/radicle-dev/radicle-client-services) for true peer-to-peer
collaboration.

But, the push-pull relationship between a local machine, seed node, and collaborators means that the project's
location—which seed node contains the most up-to-date state—is a prominent part of the user experience. You *have* to
know and remember which seed node you're working with.

Radicle's peer-to-peer protocol will solve this problem.

:::info

The open-source code behind Radicle's seed nodes is maintained by the clients team at
[`radicle-client-services`](https://github.com/radicle-dev/radicle-client-services).

:::

> ## Peer-to-peer protocol

Reviewer 3: My only note here is that people who are familiar with the idea of p2p will expect devices to be able to gossip directly with each other.

Unfortunately, this problem wasn't solved in today's current state because ISPs and firewalls are a pain in the ass.

So while gossip does happen, edge nodes will only be able to directly replicate from seeds and not other edges.

Author: Can you dig into edge nodes a bit more? I haven't seen that term come up in the context of Radicle, so I'm curious about the implementation.

Reviewer 3: Ya sorry, edge nodes are generally ephemeral devices on the _edge_ of the network, i.e. laptops, phones, etc. Opposed to servers which are the always-on, inner nodes in the network.

To put my original comment another way, "our laptops can't directly connect to each other because of pesky ISP routers"
